wrapper/src/main/java/software/amazon/jdbc/targetdriverdialect/GenericTargetDriverDialect.java

```diff
@@ -0,0 +1,117 @@
+import software.amazon.jdbc.HostSpec;
+import software.amazon.jdbc.PropertyDefinition;
+import software.amazon.jdbc.util.PropertyUtils;
+
+public class GenericTargetDriverDialect implements TargetDriverDialect {
+
+  private static final Logger LOGGER =
+      Logger.getLogger(GenericTargetDriverDialect.class.getName());
+
+  @Override
+  public boolean isDialect(Driver driver) {
+    return true;
+  }
+
+  @Override
+  public boolean isDialect(String dataSourceClass) {
+    return true;
+  }
+
+  @Override
+  public ConnectInfo prepareConnectInfo(final @NonNull String protocol,
+      final @NonNull HostSpec hostSpec,
+      final @NonNull Properties props) throws SQLException {
+
+    final String databaseName =
+        PropertyDefinition.DATABASE.getString(props) != null
+            ? PropertyDefinition.DATABASE.getString(props)
+            : "";
+    String urlBuilder = protocol + hostSpec.getUrl() + databaseName;
+
+    // keep unknown properties (the ones that don't belong to AWS Wrapper Driver)
+    // and use them to make a connection
+    PropertyDefinition.removeAllExceptCredentials(props);
+    return new ConnectInfo(urlBuilder, props);
+  }
+
+  @Override
+  public void prepareDataSource(
+      final @NonNull DataSource dataSource,
+      final @NonNull String protocol,
+      final @NonNull HostSpec hostSpec,
+      final @NonNull Properties props,
+      final @Nullable String serverPropertyName,
+      final @Nullable String portPropertyName,
+      final @Nullable String urlPropertyName,
+      final @Nullable String databasePropertyName) throws SQLException {
+
+    final Properties customProps = new Properties();
+
+    if (!isNullOrEmpty(serverPropertyName)) {
+      customProps.setProperty(serverPropertyName, hostSpec.getHost());
+    }
+
+    if (hostSpec.isPortSpecified() && !isNullOrEmpty(portPropertyName)) {
+      customProps.put(portPropertyName, hostSpec.getPort());
+    }
+
+    if (!isNullOrEmpty(databasePropertyName)
+        && !isNullOrEmpty(PropertyDefinition.DATABASE.getString(props))) {
+      customProps.setProperty(databasePropertyName, PropertyDefinition.DATABASE.getString(props));
+    }
+
+    if (!isNullOrEmpty(urlPropertyName)) {
+      final Properties urlProperties = PropertyUtils.copyProperties(customProps);
+
+      if (!isNullOrEmpty(props.getProperty(urlPropertyName))) {
+        // Remove the current url property to replace with a new url built from updated HostSpec and properties
+        urlProperties.remove(urlPropertyName);
+      }
+
+      String finalUrl = buildUrl(
+          protocol,
+          hostSpec,
+          serverPropertyName,
+          portPropertyName,
+          databasePropertyName,
+          urlProperties);
+      LOGGER.finest(() -> "Connecting to " + finalUrl);
+      customProps.setProperty(urlPropertyName, finalUrl);
+    }
+
+    LOGGER.finest(() -> PropertyUtils.logProperties(customProps, "Connecting with properties: \n"));
```

Comment:
Can we log the properties and the url in one message? It helps the debugging process in multithreaded applications

Reply:
Hmm they are different properties. Url may be missed. Let's find a separate solution for multithreaded logs.